### PR TITLE
Introduce PendingScript

### PR DESCRIPTION
**Pull request by nox**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14959)

### CODE REVIEW

File: components/script/dom/document.rs [hunk 1]
```diff
@@ -1451,12 +1451,12 @@ impl Document {
     }
 
     pub fn set_pending_parsing_blocking_script(&self, script: &HTMLScriptElement) {
-        assert!(self.get_pending_parsing_blocking_script().is_none());
+        assert!(self.has_pending_parsing_blocking_script());
```
Comment (Ms2ger) [inline components/script/dom/document.rs]: You've got that backwards.

File: components/script/dom/document.rs [hunk 2]
```diff
@@ -1601,58 +1613,66 @@ impl Document {
...
+        self.deferred_scripts.loaded(element, result);
+        self.process_deferred_scripts();
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#the-end step 3.
     pub fn process_deferred_scripts(&self) {
-        if self.ready_state.get() != DocumentReadyState::Interactive {
-            return;
-        }
+        assert!(self.ready_state.get() == DocumentReadyState::Interactive);
```
Comment (Ms2ger) [inline components/script/dom/document.rs]: `assert_eq`? I also don't see why it would be true.

File: components/script/dom/document.rs [hunk 3]
```diff
@@ -1601,58 +1613,66 @@ impl Document {
...
+    /// https://html.spec.whatwg.org/multipage/#the-end step 3.
+    /// https://html.spec.whatwg.org/multipage/#prepare-a-script step 22.c.
+    pub fn asap_in_order_script_loaded(&self,
+                                       element: &HTMLScriptElement,
+                                       result: ScriptResult) {
+        self.asap_in_order_scripts_list.loaded(element, result);
+    }
+
+    fn process_asap_scripts(&self) {
+        let mut scripts = self.asap_scripts_set.borrow_mut();
```
Comment (Ms2ger) [inline components/script/dom/document.rs]: I wonder if it's possible to make this panic by doing something along the lines of
```html
<script async>
var s = document.createElement("script");
s.setAttribute("async", "");
s.src = ...;
document.appendChild(s);
</script>
```
Could you try breaking it?

File: components/script/dom/document.rs [hunk 4]
```diff
@@ -1601,58 +1613,66 @@ impl Document {
...
+        let mut blocking_script = self.pending_parsing_blocking_script.borrow_mut();
+        let entry = blocking_script.as_mut().unwrap();
+        assert!(&*entry.element == element);
+        entry.loaded(result);
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#prepare-a-script step 22.d.
+    pub fn deferred_script_loaded(&self, element: &HTMLScriptElement, result: ScriptResult) {
+        self.deferred_scripts.loaded(element, result);
+        self.process_deferred_scripts();
```
Comment (Ms2ger) [inline components/script/dom/document.rs]: This seems broken. We should not be executing those just because they loaded.

**Review by Ms2ger — CHANGES_REQUESTED** [review]

**Comment by Ms2ger** [comment]
Let's see what bors thinks

@bors-servo r+

**Comment by nox** [comment]
@bors-servo r=Ms2ger

> 17:44 &lt;nox> Ms2ger: That assertion in process_deferred_scripts was premature, can't do that as long as finish_load exists.
> 17:44 &lt;nox> Ms2ger: Reverting that part and calling process_deferred_scripts from finish_load again makes the failure go away.

**Pull request merged**